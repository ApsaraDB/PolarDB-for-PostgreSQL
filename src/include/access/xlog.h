/*
 * xlog.h
 *
 * PostgreSQL write-ahead log manager
 *
 * Portions Copyright (c) 2020, Alibaba Group Holding Limited
 * Portions Copyright (c) 1996-2018, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 * src/include/access/xlog.h
 */
#ifndef XLOG_H
#define XLOG_H

#include "access/rmgr.h"
#include "access/xlogdefs.h"
#include "access/xloginsert.h"
#include "access/xlogreader.h"
#include "datatype/timestamp.h"
#include "lib/stringinfo.h"
#include "nodes/pg_list.h"
#include "storage/fd.h"
#ifdef ENABLE_PARALLEL_RECOVERY
#include "storage/shm_toc.h"
#include "storage/dsm.h"
#endif


/* Sync methods */
#define SYNC_METHOD_FSYNC		0
#define SYNC_METHOD_FDATASYNC	1
#define SYNC_METHOD_OPEN		2	/* for O_SYNC */
#define SYNC_METHOD_FSYNC_WRITETHROUGH	3
#define SYNC_METHOD_OPEN_DSYNC	4	/* for O_DSYNC */
extern int	sync_method;

extern PGDLLIMPORT TimeLineID ThisTimeLineID;	/* current TLI */

/*
 * Prior to 8.4, all activity during recovery was carried out by the startup
 * process. This local variable continues to be used in many parts of the
 * code to indicate actions taken by RecoveryManagers. Other processes that
 * potentially perform work during recovery should check RecoveryInProgress().
 * See XLogCtl notes in xlog.c.
 */
extern bool InRecovery;

/*
 * Like InRecovery, standbyState is only valid in the startup process.
 * In all other processes it will have the value STANDBY_DISABLED (so
 * InHotStandby will read as false).
 *
 * In DISABLED state, we're performing crash recovery or hot standby was
 * disabled in postgresql.conf.
 *
 * In INITIALIZED state, we've run InitRecoveryTransactionEnvironment, but
 * we haven't yet processed a RUNNING_XACTS or shutdown-checkpoint WAL record
 * to initialize our master-transaction tracking system.
 *
 * In SNAPSHOT_READY mode, we have full knowledge of transactions that are
 * (or were) running in the master at the current WAL location. Snapshots
 * can be taken, and read-only queries can be run.
 */
typedef enum
{
	STANDBY_DISABLED,
	STANDBY_INITIALIZED,
	STANDBY_SNAPSHOT_READY
} HotStandbyState;

extern HotStandbyState standbyState;

#define InHotStandby (standbyState >= STANDBY_SNAPSHOT_READY)

/*
 * Recovery target type.
 * Only set during a Point in Time recovery, not when standby_mode = on
 */
typedef enum
{
	RECOVERY_TARGET_UNSET,
	RECOVERY_TARGET_XID,
	RECOVERY_TARGET_TIME,
	RECOVERY_TARGET_NAME,
	RECOVERY_TARGET_LSN,
	RECOVERY_TARGET_IMMEDIATE
} RecoveryTargetType;

extern XLogRecPtr ProcLastRecPtr;
extern XLogRecPtr XactLastRecEnd;
extern PGDLLIMPORT XLogRecPtr XactLastCommitEnd;

extern bool reachedConsistency;

/* these variables are GUC parameters related to XLOG */
extern int	wal_segment_size;
extern int	min_wal_size_mb;
extern int	max_wal_size_mb;
extern int	wal_keep_segments;
extern int	XLOGbuffers;
extern int	XLogArchiveTimeout;
extern int	wal_retrieve_retry_interval;
extern char *XLogArchiveCommand;
extern bool EnableHotStandby;
extern bool fullPageWrites;
extern bool wal_log_hints;
extern bool wal_compression;
extern bool wal_warmup;
extern bool *wal_consistency_checking;
extern char *wal_consistency_checking_string;
extern bool log_checkpoints;

extern int	CheckPointSegments;

#ifdef ENABLE_PARALLEL_RECOVERY
extern int	max_parallel_replay_workers;
extern bool enable_parallel_recovery_print;
extern bool enable_parallel_recovery_bypage;
extern bool enable_parallel_recovery_locklog;
extern bool AllowHotStandbyInconsistency;
extern int	max_workload_adjust_period;
extern double parallel_replay_workload_fluctuation_factor;
extern bool enable_dynamic_adjust_workload;
#endif

#ifdef ENABLE_REMOTE_RECOVERY
extern bool checkpointSyncStandby;
extern XLogRecPtr checkpointRedo;
extern TimeLineID checkpointTLI;
extern bool fullPageRemoteFetch;
#endif

/* Archive modes */
typedef enum ArchiveMode
{
	ARCHIVE_MODE_OFF = 0,		/* disabled */
	ARCHIVE_MODE_ON,			/* enabled while server is running normally */
	ARCHIVE_MODE_ALWAYS			/* enabled always (even during recovery) */
} ArchiveMode;
extern int	XLogArchiveMode;

/* WAL levels */
typedef enum WalLevel
{
	WAL_LEVEL_MINIMAL = 0,
	WAL_LEVEL_REPLICA,
	WAL_LEVEL_LOGICAL
} WalLevel;

extern PGDLLIMPORT int wal_level;

/* Is WAL archiving enabled (always or only while server is running normally)? */
#define XLogArchivingActive() \
	(AssertMacro(XLogArchiveMode == ARCHIVE_MODE_OFF || wal_level >= WAL_LEVEL_REPLICA), XLogArchiveMode > ARCHIVE_MODE_OFF)
/* Is WAL archiving enabled always (even during recovery)? */
#define XLogArchivingAlways() \
	(AssertMacro(XLogArchiveMode == ARCHIVE_MODE_OFF || wal_level >= WAL_LEVEL_REPLICA), XLogArchiveMode == ARCHIVE_MODE_ALWAYS)
#define XLogArchiveCommandSet() (XLogArchiveCommand[0] != '\0')

/*
 * Is WAL-logging necessary for archival or log-shipping, or can we skip
 * WAL-logging if we fsync() the data before committing instead?
 */
#define XLogIsNeeded() (wal_level >= WAL_LEVEL_REPLICA)

/*
 * Is a full-page image needed for hint bit updates?
 *
 * Normally, we don't WAL-log hint bit updates, but if checksums are enabled,
 * we have to protect them against torn page writes.  When you only set
 * individual bits on a page, it's still consistent no matter what combination
 * of the bits make it to disk, but the checksum wouldn't match.  Also WAL-log
 * them if forced by wal_log_hints=on.
 */
#define XLogHintBitIsNeeded() (DataChecksumsEnabled() || wal_log_hints)

/* Do we need to WAL-log information required only for Hot Standby and logical replication? */
#define XLogStandbyInfoActive() (wal_level >= WAL_LEVEL_REPLICA)

/* Do we need to WAL-log information required only for logical replication? */
#define XLogLogicalInfoActive() (wal_level >= WAL_LEVEL_LOGICAL)

#ifdef WAL_DEBUG
extern bool XLOG_DEBUG;
#endif

/*
 * OR-able request flag bits for checkpoints.  The "cause" bits are used only
 * for logging purposes.  Note: the flags must be defined so that it's
 * sensible to OR together request flags arising from different requestors.
 */

/* These directly affect the behavior of CreateCheckPoint and subsidiaries */
#define CHECKPOINT_IS_SHUTDOWN	0x0001	/* Checkpoint is for shutdown */
#define CHECKPOINT_END_OF_RECOVERY	0x0002	/* Like shutdown checkpoint, but
											 * issued at end of WAL recovery */
#define CHECKPOINT_IMMEDIATE	0x0004	/* Do it without delays */
#define CHECKPOINT_FORCE		0x0008	/* Force even if no activity */
#define CHECKPOINT_FLUSH_ALL	0x0010	/* Flush all pages, including those
										 * belonging to unlogged tables */
/* These are important to RequestCheckpoint */
#define CHECKPOINT_WAIT			0x0020	/* Wait for completion */
/* These indicate the cause of a checkpoint request */
#define CHECKPOINT_CAUSE_XLOG	0x0040	/* XLOG consumption */
#define CHECKPOINT_CAUSE_TIME	0x0080	/* Elapsed time */

/*
 * Flag bits for the record being inserted, set using XLogSetRecordFlags().
 */
#define XLOG_INCLUDE_ORIGIN		0x01	/* include the replication origin */
#define XLOG_MARK_UNIMPORTANT	0x02	/* record not important for durability */


/* Checkpoint statistics */
typedef struct CheckpointStatsData
{
	TimestampTz ckpt_start_t;	/* start of checkpoint */
	TimestampTz ckpt_write_t;	/* start of flushing buffers */
	TimestampTz ckpt_sync_t;	/* start of fsyncs */
	TimestampTz ckpt_sync_end_t;	/* end of fsyncs */
	TimestampTz ckpt_end_t;		/* end of checkpoint */

	int			ckpt_bufs_written;	/* # of buffers written */

	int			ckpt_segs_added;	/* # of new xlog segments created */
	int			ckpt_segs_removed;	/* # of xlog segments deleted */
	int			ckpt_segs_recycled; /* # of xlog segments recycled */

	int			ckpt_sync_rels; /* # of relations synced */
	uint64		ckpt_longest_sync;	/* Longest sync for one relation */
	uint64		ckpt_agg_sync_time; /* The sum of all the individual sync
									 * times, which is not necessarily the
									 * same as the total elapsed time for the
									 * entire sync phase. */
} CheckpointStatsData;

extern CheckpointStatsData CheckpointStats;

struct XLogRecData;

extern XLogRecPtr XLogInsertRecord(struct XLogRecData *rdata,
				 XLogRecPtr fpw_lsn,
				 uint8 flags);
extern void XLogFlush(XLogRecPtr RecPtr);
extern bool XLogBackgroundFlush(void);
extern bool XLogNeedsFlush(XLogRecPtr RecPtr);
extern int	XLogFileInit(XLogSegNo segno, bool *use_existent, bool use_lock);
extern int	XLogFileOpen(XLogSegNo segno);

extern void CheckXLogRemoved(XLogSegNo segno, TimeLineID tli);
extern XLogSegNo XLogGetLastRemovedSegno(void);
extern void XLogSetAsyncXactLSN(XLogRecPtr record);
extern void XLogSetReplicationSlotMinimumLSN(XLogRecPtr lsn);

extern void xlog_redo(XLogReaderState *record);
extern void xlog_desc(StringInfo buf, XLogReaderState *record);
extern const char *xlog_identify(uint8 info);

extern void issue_xlog_fsync(int fd, XLogSegNo segno);

extern bool RecoveryInProgress(void);
extern bool HotStandbyActive(void);
extern bool HotStandbyActiveInReplay(void);
extern bool XLogInsertAllowed(void);
extern void GetXLogReceiptTime(TimestampTz *rtime, bool *fromStream);
extern XLogRecPtr GetXLogReplayRecPtr(TimeLineID *replayTLI);
extern XLogRecPtr GetXLogInsertRecPtr(void);
extern XLogRecPtr GetXLogWriteRecPtr(void);
extern bool RecoveryIsPaused(void);
extern void SetRecoveryPause(bool recoveryPause);
extern TimestampTz GetLatestXTime(void);
extern TimestampTz GetCurrentChunkReplayStartTime(void);
extern char *XLogFileNameP(TimeLineID tli, XLogSegNo segno);

extern void UpdateControlFile(void);
extern uint64 GetSystemIdentifier(void);
extern char *GetMockAuthenticationNonce(void);
extern bool DataChecksumsEnabled(void);
extern XLogRecPtr GetFakeLSNForUnloggedRel(void);
extern Size XLOGShmemSize(void);
extern void XLOGShmemInit(void);
#ifdef ENABLE_PARALLEL_RECOVERY
extern void ParallelRecoveryInit(void);
extern Size ParallelRecoveryShmemSize(void);
extern void ParallelRecoveryStart(const char *standbyConnInfo, XLogRecPtr checkpointRedo, TimeLineID checkpointTLI);
extern void ParallelRecoveryFinish(void);
extern void ParallelRecoveryWorkerMain(Datum main_arg);
extern bool DispatchWalRecord(XLogRecord *record, XLogReaderState *xlogreader);
extern void WaitForWorkersSyncDone(void);
extern bool IsBlockAssignedToThisWorker(Oid relFile, ForkNumber forknum, BlockNumber blkno);
extern int	GetParallelRedoWorkerId(void);

extern int	max_queue_size;
extern int	replay_buffer_size;
#endif
extern void BootStrapXLOG(void);
extern void LocalProcessControlFile(bool reset);
extern void StartupXLOG(void);
extern void ShutdownXLOG(int code, Datum arg);
extern void InitXLOGAccess(void);
extern void CreateCheckPoint(int flags);
extern bool CreateRestartPoint(int flags);
extern void XLogPutNextOid(Oid nextOid);
extern XLogRecPtr XLogRestorePoint(const char *rpName);
extern void UpdateFullPageWrites(void);
extern void GetFullPageWriteInfo(XLogRecPtr *RedoRecPtr_p, bool *doPageWrites_p);
extern XLogRecPtr GetRedoRecPtr(void);
extern XLogRecPtr GetInsertRecPtr(void);
extern XLogRecPtr GetFlushRecPtr(void);
extern XLogRecPtr GetLastImportantRecPtr(void);
extern void GetNextXidAndEpoch(TransactionId *xid, uint32 *epoch);
extern void RemovePromoteSignalFiles(void);

extern bool CheckPromoteSignal(void);
extern void WakeupRecovery(void);
extern void SetWalWriterSleeping(bool sleeping);

extern void XLogRequestWalReceiverReply(void);

extern void assign_max_wal_size(int newval, void *extra);
extern void assign_checkpoint_completion_target(double newval, void *extra);

/*
 * Routines to start, stop, and get status of a base backup.
 */

/*
 * Session-level status of base backups
 *
 * This is used in parallel with the shared memory status to control parallel
 * execution of base backup functions for a given session, be it a backend
 * dedicated to replication or a normal backend connected to a database. The
 * update of the session-level status happens at the same time as the shared
 * memory counters to keep a consistent global and local state of the backups
 * running.
 */
typedef enum SessionBackupState
{
	SESSION_BACKUP_NONE,
	SESSION_BACKUP_EXCLUSIVE,
	SESSION_BACKUP_NON_EXCLUSIVE
} SessionBackupState;

extern XLogRecPtr do_pg_start_backup(const char *backupidstr, bool fast,
				   TimeLineID *starttli_p, StringInfo labelfile,
				   List **tablespaces, StringInfo tblspcmapfile, bool infotbssize,
				   bool needtblspcmapfile);
extern XLogRecPtr do_pg_stop_backup(char *labelfile, bool waitforarchive,
				  TimeLineID *stoptli_p);
extern void do_pg_abort_backup(void);
extern SessionBackupState get_backup_status(void);
int			GetSyncBit(void);

#ifdef ENABLE_PARALLEL_RECOVERY
typedef struct
{
	RelFileNode key;			/* lookup key - must be first */
	int			workerIndex;
	int			msgCnt;
}			ParallelRedoRelfilenodeMapEntry;

extern HTAB *ParallelRedoRelfilenodeMapHash;
#endif

extern XLogRecPtr XLogGetReplicationSlotMinimumLSN(void);
extern void polar_update_last_removed_ptr(char *filename);

/* POLAR Consensus */
extern bool polar_check_pm_in_state_change(void);
extern void polar_signal_pm_state_change(int state,
							 const char *leaderAddr, int leaderPort, uint64 term,
							 uint64 nextAppendTerm, uint32 tli, uint64 logUpto);
extern void polar_signal_recovery_state_change(
								   bool newLeader, bool resumeLeader);
extern bool polar_dma_check_logger_status(char **primaryConnInfo,
							  XLogRecPtr *receivedUpto, TimeLineID *receivedTLI,
							  bool *requestNextTLI);
extern int polar_wait_recovery_wakeup(int wakeEvents, long timeout,
						   uint32 wait_event_info);
extern bool polar_is_dma_data_node(void);
extern bool polar_is_dma_logger_node(void);

/* File path names (all relative to $PGDATA) */
#define BACKUP_LABEL_FILE		"backup_label"
#define BACKUP_LABEL_OLD		"backup_label.old"

#define TABLESPACE_MAP			"tablespace_map"
#define TABLESPACE_MAP_OLD		"tablespace_map.old"

#endif							/* XLOG_H */
