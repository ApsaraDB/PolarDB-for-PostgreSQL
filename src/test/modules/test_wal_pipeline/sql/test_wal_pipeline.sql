CREATE EXTENSION test_wal_pipeline;

SELECT test_wal_pipeline();