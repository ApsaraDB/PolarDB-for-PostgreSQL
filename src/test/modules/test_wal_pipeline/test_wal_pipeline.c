#include "postgres.h"
#include "pgstat.h"
#include "utils/elog.h"
#include "utils/guc.h"
#include "miscadmin.h"
#include "storage/lwlock.h"
#include "storage/proc.h"
#include "storage/spin.h"
#include "access/xlog.h"
#include "catalog/pg_control.h"
    
PG_MODULE_MAGIC;

pthread_t	advance_handle;
pthread_t	write_handle;
pthread_t	flush_handle;
pthread_t	notify_handle;
int         ident = 0;

PG_FUNCTION_INFO_V1(test_wal_pipeline);

static void *
test_worker_for_advance(void *arg)
{
	pg_usleep(1000000); /* wait 1s */
	Assert(MyProc->wait_event_info == WAIT_EVENT_WAL_PIPELINE_WAIT_RECENT_WRITTEN_SPACE);
	polar_wal_pipeline_advance(0);
	pg_usleep(1000000); /* wait 1s */
	Assert(MyProc->wait_event_info == 0);

	pthread_exit(0);
}

/*
static void *
test_worker_for_write(void *arg)
{
	pg_usleep(1000000);
	polar_wal_pipeline_write(0);
	polar_wal_pipeline_flush(0);
	polar_wal_pipeline_notify(0);

	pthread_exit(0);
}
*/

static void
test_wal_pipeline_advance(void)
{
	elog(INFO, "------------------------------");
	elog(INFO, "%s", __FUNCTION__);

	elog(INFO, "case 1 test recovery mode");
	polar_wal_pipeline_set_local_recovery_mode(true);
	Assert(!polar_wal_pipeline_advance(0));
	polar_wal_pipeline_set_local_recovery_mode(false);
	elog(INFO, "case 1 test recovery mode ok");

	polar_wal_pipeline_set_ready_write_lsn(40);

	elog(INFO, "case 2 test add link");
	polar_wal_pipeline_recent_written_add_link(40, 60);
	polar_wal_pipeline_advance(0);
	Assert(polar_wal_pipeline_get_ready_write_lsn() == 60);
	elog(INFO, "case 2 test add link ok");

	elog(INFO, "case 3 test no space");
	polar_wal_pipeline_recent_written_add_link(60, 1084);
	Assert(MyProc->wait_event_info == 0);
	pthread_create(&advance_handle, NULL, test_worker_for_advance, &ident);
 	/* polar_wal_pipeline_recent_written_array_size is 1024 */
 	polar_wal_pipeline_recent_written_add_link(1084, 1094);   
	pthread_join(advance_handle, NULL);
	polar_wal_pipeline_advance(0);
	Assert(polar_wal_pipeline_get_ready_write_lsn() == 1094);
	elog(INFO, "case 3 test no space ok");
}

static void
test_wal_pipeline_write(void)
{
//	XLogRecPtr lsn;
//	char data[1048576] = {0};

	elog(INFO, "------------------------------");
	elog(INFO, "%s", __FUNCTION__);
	
	elog(INFO, "case 1 test recovery mode");
	polar_wal_pipeline_set_local_recovery_mode(true);
	Assert(!polar_wal_pipeline_write(0));
	polar_wal_pipeline_set_local_recovery_mode(false);
	elog(INFO, "case 1 test recovery mode ok");

	elog(INFO, "case 2 test write nothing");
	polar_wal_pipeline_set_ready_write_lsn(polar_wal_pipeline_get_write_lsn()-1);
	Assert(!polar_wal_pipeline_write(0));
	elog(INFO, "case 2 test write nothing ok");

	/*
	elog(INFO, "case 3 test write only");
	XLogBeginInsert();
	XLogRegisterData(data, sizeof(data));
	lsn = XLogInsert(RM_XLOG_ID, XLOG_NOOP);
	polar_wal_pipeline_set_ready_write_lsn(lsn);
	polar_wal_pipeline_write(0);
	Assert(polar_wal_pipeline_get_write_lsn() == lsn);
	elog(INFO, "case 3 test write only ok");

	elog(INFO, "case 4 test flush only");
	pthread_create(&write_handle, NULL, test_worker_for_write, &ident);  
	polar_wal_pipeline_commit_wait(lsn);
	pthread_join(write_handle, NULL);
	Assert(polar_wal_pipeline_get_flush_lsn() == lsn);
	elog(INFO, "case 4 test flush only ok");
	*/
}

static void
test_wal_pipeline_flush(void)
{
	elog(INFO, "------------------------------");
	elog(INFO, "%s", __FUNCTION__);

	elog(INFO, "case 1 test recovery mode");
	polar_wal_pipeline_set_local_recovery_mode(true);
	Assert(!polar_wal_pipeline_flush(0));
	polar_wal_pipeline_set_local_recovery_mode(false);
	elog(INFO, "case 1 test recovery mode ok");
}

static void
test_wal_pipeline_notify(void)
{
	elog(INFO, "------------------------------");
	elog(INFO, "%s", __FUNCTION__);

	elog(INFO, "case 1 test recovery mode");
	polar_wal_pipeline_set_local_recovery_mode(true);
//	Assert(!polar_wal_pipeline_notify(0));
	polar_wal_pipeline_set_local_recovery_mode(false);
	elog(INFO, "case 1 test recovery mode ok");
}

/*
 * SQL-callable entry point to perform all tests.
 */
Datum
test_wal_pipeline(PG_FUNCTION_ARGS)
{
	test_wal_pipeline_advance();
	
	test_wal_pipeline_write();

	test_wal_pipeline_flush();

	test_wal_pipeline_notify();

	PG_RETURN_VOID();
}
