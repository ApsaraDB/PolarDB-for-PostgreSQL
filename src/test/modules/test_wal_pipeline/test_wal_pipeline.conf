polar_wal_pipeline_enable = false
polar_wal_pipeline_recent_written_array_size = 1024
# no checkpoint and walwriter
checkpoint_timeout = 86400
wal_writer_delay = 10000
