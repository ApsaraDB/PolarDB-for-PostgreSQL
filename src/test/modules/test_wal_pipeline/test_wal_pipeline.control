comment = 'Test code for polar wal pipeline'
default_version = '1.0'
module_pathname = '$libdir/test_wal_pipeline'
relocatable = true