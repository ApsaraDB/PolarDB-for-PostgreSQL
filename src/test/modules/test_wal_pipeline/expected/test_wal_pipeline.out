CREATE EXTENSION test_wal_pipeline;
SELECT test_wal_pipeline();
INFO:  ------------------------------
INFO:  test_wal_pipeline_advance
INFO:  case 1 test recovery mode
INFO:  case 1 test recovery mode ok
INFO:  case 2 test add link
INFO:  case 2 test add link ok
INFO:  case 3 test no space
INFO:  case 3 test no space ok
INFO:  ------------------------------
INFO:  test_wal_pipeline_write
INFO:  case 1 test recovery mode
INFO:  case 1 test recovery mode ok
INFO:  case 2 test write nothing
INFO:  case 2 test write nothing ok
INFO:  ------------------------------
INFO:  test_wal_pipeline_flush
INFO:  case 1 test recovery mode
INFO:  case 1 test recovery mode ok
INFO:  ------------------------------
INFO:  test_wal_pipeline_notify
INFO:  case 1 test recovery mode
INFO:  case 1 test recovery mode ok
 test_wal_pipeline 
-------------------
 
(1 row)

